### DISCUSSION TITLE
Update from 11 to 17 and my notifications stopped working.

### DISCUSSION THREAD

**Discussion by torbentschechne** [dpa99c/cordova-plugin-firebasex#904] [Q&A] [answered]
Hey there,

I updated `cordova-android` from 9 to 13 and `cordova-plugin-firebasex` from 11.0.3-cli to 17.0.0.

Then I went to the `package.json` and updated the versions as shown here (https://github.com/dpa99c/cordova-plugin-firebasex/blob/master/plugin.xml#L80) to this:

```
"cordova-plugin-firebasex": {
                "FIREBASE_ANALYTICS_COLLECTION_ENABLED": "true",
                "FIREBASE_PERFORMANCE_COLLECTION_ENABLED": "true",
                "FIREBASE_CRASHLYTICS_COLLECTION_ENABLED": "true",
                "FIREBASE_FCM_AUTOINIT_ENABLED": "true",
                "FIREBASE_ANALYTICS_WITHOUT_ADS": "false",
                "GOOGLE_ANALYTICS_ADID_COLLECTION_ENABLED": "true",
                "GOOGLE_ANALYTICS_DEFAULT_ALLOW_ANALYTICS_STORAGE": "true",
                "GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_STORAGE": "true",
                "GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_USER_DATA": "true",
                "GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_PERSONALIZATION_SIGNALS": "true",
                "ANDROID_ICON_ACCENT": "#FF00FFFF",
                "ANDROID_FIREBASE_PERFORMANCE_MONITORING": "false",
                "ANDROID_PLAY_SERVICES_TAGMANAGER_VERSION": "18.1.0",
                "ANDROID_PLAY_SERVICES_AUTH_VERSION": "21.2.0",
                "ANDROID_FIREBASE_ANALYTICS_VERSION": "22.1.0",
                "ANDROID_FIREBASE_MESSAGING_VERSION": "24.0.1",
                "ANDROID_FIREBASE_CONFIG_VERSION": "22.0.0",
                "ANDROID_FIREBASE_PERF_VERSION": "21.0.1",
                "ANDROID_FIREBASE_AUTH_VERSION": "23.0.0",
                "ANDROID_FIREBASE_INAPPMESSAGING_VERSION": "21.0.0",
                "ANDROID_FIREBASE_FIRESTORE_VERSION": "25.1.0",
                "ANDROID_FIREBASE_FUNCTIONS_VERSION": "21.0.0",
                "ANDROID_FIREBASE_IID_VERSION": "21.1.0",
                "ANDROID_FIREBASE_INSTALLATIONS_VERSION": "18.0.0",
                "ANDROID_FIREBASE_CRASHLYTICS_VERSION": "19.1.0",
                "ANDROID_FIREBASE_CRASHLYTICS_NDK_VERSION": "19.1.0",
                "ANDROID_GSON_VERSION": "2.11.0",
                "ANDROID_FIREBASE_PERF_GRADLE_PLUGIN_VERSION": "1.4.2",
                "ANDROID_GRPC_OKHTTP": "1.68.0"
            }
```

Then I reinstalled the plugin `cordova plugin rm cordova-plugin-firebasex --nosave && cordova plugin add cordova-plugin-firebasex --nosave`

I also downloaded the `google-services.json` and `GoogleService-Info.plist` and replaced the downloaded ones with the ones in the project. After setting the right JDK and Gradle versions, I compiled the app.

However, the in app notifications are not working. I opened the app and went to the home screen. Then I simulated to send me a message, but no notification is popping up. I am not that familiar with Cordova/Android development and now questioning myself: What else could I've forgotten? Can anybody help?

**Comment by dpa99c** [comment]
Hi @torbentschechne, you can try building the [example project](https://github.com/dpa99c/cordova-plugin-firebasex-test) and use its built-in messaging client to send a test message as described in the documentation.
Note that you'll need to change the package ID in `config.xml` to one which you have a `google-services.json` for, before adding the android platform.

**Reply by torbentschechne** [reply]
Thanks for your quick reply. You mean, I can just clone the example repo, change the package ID, send the notification in my application and can see if the notification pops up?

When I look through the `config.xml` of your example, I see that you are doing this: 

```
<config-file target="AndroidManifest.xml" parent="/*">
   <uses-permission android:name="android.permission.POST_NOTIFICATIONS" />
</config-file>
```

In my config.xml I have this one:

```
<edit-config file="app/src/main/AndroidManifest.xml" mode="merge" target="/manifest/application">
    <application xmlns:android="http://schemas.android.com/apk/res/android" android:usesCleartextTraffic="true" />
</edit-config>
```

The tag is not `config-file` but `edit-config` and has a different contents. Should I replace it with yours?

Also when I put some error logs into AI, I get a result like this:

```
In the non-working log, push_enabled is explicitly set to false in settings retrieved by SessionConfigFetcher. This may be why notifications aren’t being sent. In contrast, the working log lacks this setting, suggesting push notifications may be enabled by default or via other configurations.
```

```
Push Setting Check: Explicitly check or enable push_enabled in the app's Firebase configuration, as this setting might now control push notifications.
```

Do you know where I can set this or should I ignore it?

**Reply by dpa99c** [reply]
> Thanks for your quick reply. You mean, I can just clone the example repo, change the package ID, send the notification in my application and can see if the notification pops up?
> 
> When I look through the `config.xml` of your example, I see that you are doing this:
> 
> ```
> <config-file target="AndroidManifest.xml" parent="/*">
>    <uses-permission android:name="android.permission.POST_NOTIFICATIONS" />
> </config-file>
> ```

`POST_NOTIFICATIONS` is a run-time permission required on Android 13 (API level 33) and above.
If your app is built with API 33+ SDK and is running on an device with Android 13 (API level 33) or above, you need to add this permission to your `AndroidManifest.xml` using the above block and request the permission at run-time e.g. [grantPermission()](https://github.com/dpa99c/cordova-plugin-firebasex?tab=readme-ov-file#grantpermission), otherwise your app will not receive notifications.

> 
> In my config.xml I have this one:
> 
> ```
> <edit-config file="app/src/main/AndroidManifest.xml" mode="merge" target="/manifest/application">
>     <application xmlns:android="http://schemas.android.com/apk/res/android" android:usesCleartextTraffic="true" />
> </edit-config>
> ```
> 
> The tag is not `config-file` but `edit-config` and has a different contents. Should I replace it with yours?

This permits your app to talk to resources over HTTP protocol (which is blocked by default by Android). It's up to you whether you need this.

> 
> Also when I put some error logs into AI, I get a result like this:
> 
> ```
> In the non-working log, push_enabled is explicitly set to false in settings retrieved by SessionConfigFetcher. This may be why notifications aren’t being sent. In contrast, the working log lacks this setting, suggesting push notifications may be enabled by default or via other configurations.
> ```
> 
> ```
> Push Setting Check: Explicitly check or enable push_enabled in the app's Firebase configuration, as this setting might now control push notifications.
> ```
> 
> Do you know where I can set this or should I ignore it?

This doesn't make any sense. Good old AI.

**Reply by torbentschechne** [reply]
I did this now:

```
<config-file target="AndroidManifest.xml" parent="/*">
   <uses-permission android:name="android.permission.POST_NOTIFICATIONS" />
</config-file>
```

And I added `xmlns:android="http://schemas.android.com/apk/res/android"` it to widget 

```
<widget id="com.mypackage.app" version="1.0.9" xmlns="http://www.w3.org/ns/widgets" xmlns:cdv="http://cordova.apache.org/ns/1.0" xmlns:android="http://schemas.android.com/apk/res/android">
```

Then I've done this:

```
cordova plugin rm cordova-plugin-firebasex --nosave && cordova plugin add cordova-plugin-firebasex --nosave
```

But the AndroidManifest.xml in `platforms/android/app/src/main/AndroidManifest.xml` just shows:

```
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.WAKE_LOCK" />
```

And nothing regarding `POST_NOTIFICATIONS`. Did I miss something?

**Reply by dpa99c** [reply]
Remove and re-add the Android platform, not the plugin

**Reply by torbentschechne** [reply]
Thanks! I did that, now it's showing up in the `AndroidManifest.xml`. I built the .aab file and uploaded it to Browserstack. The app started and asked me if I want to allow notifications which i approved, I opened the home screen of Android and I've sent a chat message through my browser (usually the phone should then display a notification), but nothing. When I look into the logs and search for firebase I can see:

```
11-06 17:00:10.446 I/FirebaseApp(19617): Device unlocked: initializing all Firebase APIs for app [DEFAULT]
11-06 17:00:10.498 I/FirebaseCrashlytics(19617): Initializing Firebase Crashlytics 19.1.0 for com.mypackage.app
11-06 17:00:10.534 I/FirebaseInitProvider(19617): FirebaseApp initialization successful
11-06 17:00:10.541 I/FirebaseCrashlytics(19617): Saved version control info
11-06 17:00:10.548 I/FA      (19617):   adb shell setprop debug.firebase.analytics.app com.mypackage.app
11-06 17:00:10.692 I/FirebasePlugin(19617): Creating channel id=fcm_default_channel
11-06 17:01:03.255 I/TRuntime.CctTransportBackend(19617): Making request to: https://firebaselogging-pa.googleapis.com/v1/firelog/legacy/batchlog
11-06 17:04:37.888 I/TRuntime.CctTransportBackend(19617): Making request to: https://firebaselogging-pa.googleapis.com/v1/firelog/legacy/batchlog
```

That looks okay, doesn't it? But I believe I saw the same logs without `android.permission.POST_NOTIFICATIONS`

Also I get some errors/warnings, like these:

```
11-06 17:20:18.087 E/com.mypackage.app(27001): Not starting debugger since process cannot load the jdwp agent.
11-06 17:20:18.252 E/com.mypackage.app(27001): No package ID 77 found for ID 0x770b000f.
11-06 17:20:18.252 E/com.mypackage.app(27001): No package ID 77 found for ID 0x770b000f.
11-06 17:20:18.738 E/GoogleApiManager(27001): Failed to get service from broker.
11-06 17:20:18.738 E/GoogleApiManager(27001): java.lang.SecurityException: Unknown calling package name 'com.google.android.gms'.
11-06 17:20:18.738 E/GoogleApiManager(27001):     at android.os.Parcel.createExceptionOrNull(Parcel.java:3011)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at android.os.Parcel.createException(Parcel.java:2995)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at android.os.Parcel.readException(Parcel.java:2978)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at android.os.Parcel.readException(Parcel.java:2920)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at angb.a(:com.google.android.gms@244336029@24.43.36 (190400-691491885):36)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at anei.z(:com.google.android.gms@244336029@24.43.36 (190400-691491885):143)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at amln.run(:com.google.android.gms@244336029@24.43.36 (190400-691491885):54)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at android.os.Handler.handleCallback(Handler.java:942)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at android.os.Handler.dispatchMessage(Handler.java:99)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at bsuh.mC(:com.google.android.gms@244336029@24.43.36 (190400-691491885):1)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at bsuh.dispatchMessage(:com.google.android.gms@244336029@24.43.36 (190400-691491885):5)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at android.os.Looper.loopOnce(Looper.java:201)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at android.os.Looper.loop(Looper.java:288)
11-06 17:20:18.738 E/GoogleApiManager(27001):     at android.os.HandlerThread.run(HandlerThread.java:67)
11-06 17:20:18.745 E/SystemWebViewClient(27001): www/favicon.ico
11-06 17:20:28.730 E/GoogleApiManager(27001): Failed to get service from broker.
11-06 17:20:28.730 E/GoogleApiManager(27001): java.lang.SecurityException: Unknown calling package name 'com.google.android.gms'.
11-06 17:20:28.730 E/GoogleApiManager(27001):     at android.os.Parcel.createExceptionOrNull(Parcel.java:3011)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at android.os.Parcel.createException(Parcel.java:2995)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at android.os.Parcel.readException(Parcel.java:2978)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at android.os.Parcel.readException(Parcel.java:2920)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at angb.a(:com.google.android.gms@244336029@24.43.36 (190400-691491885):36)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at anei.z(:com.google.android.gms@244336029@24.43.36 (190400-691491885):143)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at amln.run(:com.google.android.gms@244336029@24.43.36 (190400-691491885):54)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at android.os.Handler.handleCallback(Handler.java:942)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at android.os.Handler.dispatchMessage(Handler.java:99)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at bsuh.mC(:com.google.android.gms@244336029@24.43.36 (190400-691491885):1)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at bsuh.dispatchMessage(:com.google.android.gms@244336029@24.43.36 (190400-691491885):5)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at android.os.Looper.loopOnce(Looper.java:201)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at android.os.Looper.loop(Looper.java:288)
11-06 17:20:28.730 E/GoogleApiManager(27001):     at android.os.HandlerThread.run(HandlerThread.java:67)

```

Do you have any ideas left what I could do?

**Reply by torbentschechne** [reply]
@dpa99c I have another question and hope you could help out. The project I've taken over has this structure:

![CleanShot 2024-11-07 at 09 25 48@2x](https://github.com/user-attachments/assets/f6dc9566-e5a9-437b-becf-140dd3334a2b)

The `notification.js` starts with:

![CleanShot 2024-11-07 at 09 26 15@2x](https://github.com/user-attachments/assets/a99185d9-10b3-4cd0-ac99-745c5a6165bf)

I don't see what role these .js files play when compiling. When I set up a new cordova android project and add your firebase plugin, I do not have these `src/*.js` files. So now I wonder, if these are even necessary or are those legacy files?

**Reply by dpa99c** [reply]
Lots of Cordova projects introduce their own build steps to pre-process code before it's deployed to the `www/` directory that Cordova uses for its web assets. It's common the call this `src/` as other frameworks use this structure.
Have a look in `package.json` for build script definitions, otherwise for a `scripts/` or `bin/` directory containing the build scripts.
Maybe they've even documented this in the `README.md` - that would be nice! 😃

**Reply by torbentschechne** [reply]
Yep, the `app.js` (which includes the other .js files) is mentioned in the `webpack.config.js` which is then bundled into a `main.js`.

Do you have another idea why notifications are not working currently (maybe it's connected to these errors?): https://github.com/dpa99c/cordova-plugin-firebasex/discussions/904#discussioncomment-11169130

If this kind of support (which I highly appreciate) is beyond your scope, tell me. Maybe I could book a screen sharing session with you to address the issue?

**Reply by dpa99c** [reply]
I would use the [example project](https://github.com/dpa99c/cordova-plugin-firebasex-test) and use its built-in messaging client to send a test message as described in the documentation. This is known to be a working codebase so will validate your Firebase configuration and dev environment if you can successfully receive messages this way. 
If that works, then try sending the messages from your web console to validate that.
Then if that works, you know the problem is your own app codebase.

**Reply by torbentschechne** [reply]
I'll try that. I believe it has something to do with the updated packages, because an older version of the app (with `cordova-android` 9 and `cordova-plugin-firebasex` 11.0.3.-cli was working. I'll take a look into the example project again.

**Reply by torbentschechne** [reply]
I've done that. And it worked. I noticed this one in the docs: 

```
Press "Generate new private key" and save the file as service-account.json in the root of this cloned project repo.
```

After I've copied the file in the root of the test project, I've re-installed the firebasex plugin and removed and added android as a platform as well. Then I built the app. I noticed, that I don't have the `service-account.json` in my legacy project. I only have the `google-services.json` and the `GoogleService-Info.plist`.

1. Could that be a missing piece, is this something that came into the codebase later? I don't believe so, because you just need the service-account.json to send a message, not to receive one. Right?
2. Also interesting is, that I have a webpack.config.js, but you don't :)
3. What is currently unclear to me, why my project has a `notification.js` at all. Maybe this is not needed with the new version?
4. Maybe the final solution (if the above did not help) is to use the test project as a starting point and try to build a new app on top of that...

**Reply by dpa99c** [reply]
> 1. Could that be a missing piece, is this something that came into the codebase later?

No, `service-account.json` is only needed for the built-in messaging client in the test project. This is not something you'd use in production. In production, you'd use the Firebase Console or some backend system to send the notifications to user's devices.

> 2. Another step could be now, that I remove all the *.js scripts from my legacy project and just place the `sendMessage.js` from the test project into it, right?

`sendMessage.js` is the messaging client that uses the Firebase Admin SDK on your dev machine to send notifications (instead of using the Firebase Console, etc.). You don't need to copy it to your project: if you want to test your app using it, so long as it has the same package ID as the test project app, you can run it from the test project. 
You just need to find the FCM token assign to the app installation when your app starts up (can be seen in logcat output in Android Studio or by inspecting Webview of running app with Chrome Devtools and executing `FirebasePlugin.getToken(function(token){console.log(token);});`

You can also try it the other way round: try sending messages to the test project app using whatever other mechanism your app uses (e.g. Firebase Console).

> 3. Also interesting is, that I have a webpack.config.js, but you don't :)

Webpack is simply a build tool for packaging up the source files in your `src/` directory and deploying to `www/`.
The devs of your app project decided to use it the Cordova project.
Whereas the test project is very simple so doesn't need a code packaging tool.

> 4. Maybe the final solution (if the above did not help) is to use the test project as a starting point and try to build a new app on top of that...

The test project is just that: an example of how the plugin can be used. I would not attempt to build a production project on top of it. Particularly since the existing project has a custom webpack-based build system in place. 

`cordova-android@9` builds with API 29 whereas `cordova-android@13` builds with API 34 by default.
There are a lot of breaking changes Google made in between so review the changelogs for Android
Also lots of change to the plugin between v11 and v17 (btw I just released v18).
The problem is most likely in the app code somewhere, either build configuration, or run-time code that needs updating for the above.
You can compare the working test project to your not working app project to see what is different.

I can't really help more than that without seeing the code myself, in which case I may as well fix it myself, which is what people pay me to do.

**Reply by torbentschechne** [reply]
Got it. Thanks!

I have two questions left:

1. There isn't so much application code here. I only have the `src/*.js` files here in the Cordova project which are taken into the build process through webpack. What I just did is, that I built the app with `cordova run android` to see if these *.js files have any value. But also with this version, no notification is popping up.

2. With the API changes you only refer to the Cordova project, right? The backend has a package `firebase` which is version 9.10.0. Do you mean I should update this one as well to ~11.0.1, because the js lib that the backend uses to dispatch the messages could be incompatible to the Cordova packages?

3. I'll try to use `sendMessage.js` with my legacy project.

4. Do you have an email address where I can get a quote from you in case I fail? :)

**Reply by torbentschechne** [reply]
Interesting. When I do 3, use `sendMessage.js` in my legacy project, I do see a notification. So it can be, that 2. is valid, right?

**Reply by dpa99c** [reply]
If your legacy app received a message using the test project's messaging client, then there's nothing wrong with it.
And therefore the problem lies with how you are sending the notifications, not receiving them.

**Reply by torbentschechne** [reply]
It's still weird, I can't figure out why the sending would be the issue if the older app shows messages and the new app doesn't. Because if the old app shows notifications, it means that data must be in firebase (from the sender).

Would it be possible that you take a - of course paid - look at it?

**Reply by torbentschechne** [reply]
Hey @dpa99c, do you have an answer for me?

**Reply by dpa99c** [reply]
Sorry, forgot to reply as been busy with other stuff.

So to clarify:
- legacy app build (cordova-android@9, firebasex@11) receives notifications
- new app build (cordova-android@13, FirebaseX@17) doesn't

That correct? If so, problem is app-side, not server-side 

I may have time to take a look later this week if you still need me to

**Reply by torbentschechne** [reply]
Yep, exactly. But the new app build does receive notifications when I trigger a notification with your "sendMessage" script. I wanted to build the legacy app today from my mac (because the current app build of the legacy app is from another computer), as I wanted to see if it may be related to my mac. But here I struggle, because I always get

```
A problem occurred evaluating script.
> No usable Android build tools found. Highest 34.x installed version is 29.0.3; Recommended version is 34.0.0.
```
Independent on the `<preference name="android-targetSdkVersion" value="29.0.3" />` or `<preference name="android-compileSdkVersion" value="29.0.3" />` I also downgraded the cordova-cli to 11 and then to 10, but I have no idea why the process needs 34 ;-)

It would be great if you can take a look at it this week. How can I contact you besides here, because I would need some details from you (hourly rate etc.) to see if that makes sense. ;-)

**Reply by torbentschechne** [reply]
Hey @dpa99c,

I was able to build different versions today, and the update from 

- `cordova-android: 9.0.0` to `"cordova-android": "^10.1.2"`
- `"cordova-plugin-firebasex": "^11.0.3-cli"` to `"cordova-plugin-firebasex": "^12.1.0",`

```
<preference name="android-targetSdkVersion" value="30" />
<preference name="GradleVersion" value="6.9" />
```

worked fine. Also the update from worked fine and notifications were displayed on the home screen:

- `"cordova-android": "^10.1.2"` to `"cordova-android": "^11.0.0"`
- `"cordova-plugin-firebasex": "^12.1.0"` to `"cordova-plugin-firebasex": "^14.0.0"`,

```
<preference name="android-targetSdkVersion" value="32" />
```

Then I updated

- `"cordova-android": "^11.0.0"` to `"cordova-android": "^12.0.1"`
- `"cordova-plugin-firebasex": "^14.0.0"` to `"cordova-plugin-firebasex": "^15.0.0"`, (I also tried 16.0.0)

with 

```
<preference name="android-targetSdkVersion" value="33" />
<preference name="GradleVersion" value="7.6.4" />
<preference name="GradlePluginGoogleServicesEnabled" value="true" />
```

and

```
<config-file target="AndroidManifest.xml" parent="/*">
    <uses-permission android:name="android.permission.POST_NOTIFICATIONS" />
</config-file>
```

and since then, I do not get Notifications anymore on an Android 13 and 15. So somewhere there must be the issue, but I don't see anything in the changelogs which can cause this or if I missed something in `config.xml`. So it would be really helpful if you could take a look at this in this week.

**Reply by dpa99c** [reply, marked as answer]
Hi Torben,

I had some urgent stuff today but now it's done.
Drop me a email on dpa99c@gmail.com
You can Google Chat with me there or we can work out another way. I'm in the UK so my timezone is GMT. Let's take it from there but I'm hopeful I can help you out.
